controllers/metal3.io/baremetalhost_controller.go

```diff
@@ -1153,6 +1180,245 @@ func saveHostProvisioningSettings(host *metal3v1alpha1.BareMetalHost) (dirty boo
+
+		if err = r.updateHostFirmwareSettingsFromCurrent(currentSettings, hfs, firmwareSchema); err != nil {
+			return errors.Wrap(err, "could not update hostFirmwareSettings")
+		}
+	} else {
+		if err = r.updateHostFirmwareSettingsFromCurrent(currentSettings, hfs, nil); err != nil {
+			return errors.Wrap(err, "could not update hostFirmwareSettings")
+		}
+		info.log.Info("updated hostFirmwareSettings settings without schema")
+	}
+
+	return nil
+}
+
+// Create a HostFirmwareSettings resource
+func (r *BareMetalHostReconciler) newHostFirmwareSettings(info *reconcileInfo, hfs *metal3v1alpha1.HostFirmwareSettings) (err error) {
+
+	// Use same name and namespace as host
+	hfs.ObjectMeta = metav1.ObjectMeta{
+		Name:      info.host.ObjectMeta.Name,
+		Namespace: info.host.ObjectMeta.Namespace}
+	hfs.Status.Settings = make(metal3v1alpha1.SettingsMap)
+	hfs.Spec.Settings = make(metal3v1alpha1.DesiredSettingsMap)
+	hfs.Annotations = make(map[string]string)
+	if err = r.Create(context.TODO(), hfs); err != nil {
+		return errors.Wrap(err, "could not create hostFirmwareSettings resource")
+	}
+	info.log.Info("created new hostFirmwareSettings resource")
+
+	return nil
+}
+
+// Update the HostFirmwareSettings resource using the settings and schema from provisioner
+func (r *BareMetalHostReconciler) updateHostFirmwareSettingsFromCurrent(settings metal3v1alpha1.SettingsMap, hfs *metal3v1alpha1.HostFirmwareSettings, schema *metal3v1alpha1.FirmwareSchema) (err error) {
+
+	// if no schema provided, get the one referenced by hostFirmwareSettings if available
+	if schema == nil && hfs.Status.FirmwareSchema != nil {
+		schema = &metal3v1alpha1.FirmwareSchema{}
+		if r.Get(context.TODO(), client.ObjectKey{
+			Namespace: hfs.Status.FirmwareSchema.Namespace,
+			Name:      hfs.Status.FirmwareSchema.Name}, schema) != nil {
+			schema = nil
+		}
+	}
+
+	// If settings cleared after pivot recreate map, it will be repopulated below
+	if hfs.Status.Settings == nil {
+		hfs.Status.Settings = make(metal3v1alpha1.SettingsMap)
+	}
+
+	// Update Spec and Status
+	specDirty := false
+	for k, v := range settings {
+		// Some vendors include encrypted password fields, don't add these
+		if strings.Contains(k, "Password") {
+			continue
+		}
+		hfs.Status.Settings[k] = v
+
+		// Don't include setting in Spec if cannot be changed
+		if schema != nil {
+			if schemaSetting, ok := schema.Spec.Schema[k]; ok {
+				if (schemaSetting.ReadOnly != nil && *schemaSetting.ReadOnly == true) ||
+					(schemaSetting.Unique != nil && *schemaSetting.Unique == true) {
+					continue
+				}
+			}
+		}
+		// Copy to Spec only if doesn't exist as setting may have been set by user
+		if _, ok := hfs.Spec.Settings[k]; !ok {
+			hfs.Spec.Settings[k] = intstr.FromString(v)
+			specDirty = true
+		}
+	}
+
+	if specDirty {
+		if err = r.Update(context.TODO(), hfs); err != nil {
+			return errors.Wrap(err, "could not update host firmware settings")
+		}
+	}
+
+	return r.Status().Update(context.TODO(), hfs)
```

Comment:
Should this go to an `else` block to avoid two subsequent calls to `Update`?

Reply:
So we only need to do the full Update() when the Spec changes, the second call is just for a Status().Update(). I think we still need the Status().Update if the other Update() is done, but I will retest this.